src/context/typecore.ml

```diff
@@ -222,7 +222,31 @@ let add_local ctx k n t p =
 	ctx.locals <- PMap.add n v ctx.locals;
 	v
 
+let check_identifier_name ctx name kind p =
+	if starts_with name '$' then
+		display_error ctx (kind ^ " names starting with a dollar are not allowed") p
+	else
+		let is_valid_identifier s =
+			match s with
+			| "new" | "this" -> true
+			| _ -> Lexer.is_valid_identifier s
+		in
+		if not (is_valid_identifier name) then display_error ctx ("`" ^ name ^ "` is not a valid " ^ kind ^ " name") p
```

Comment:
You kinda have to escape backticks in `name`, right?

Reply:
You mean in case somebody renders errors as markdown? It's mostly meant as delimiters because the name could include spaces. But maybe VSCode lets us have markdown in errors sometime in the future, or some other IDE wants to do that.

Reply:
Hm, apparently escaping backticks is not so simple, it actually requires doubling the _surrounding_ backticks: https://stackoverflow.com/a/25943045/2631715

Maybe I should just use quotes.